### PR TITLE
Major refactoring

### PR DESCRIPTION
**Pull request by radito3**

### CODE REVIEW

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 1]
```diff
@@ -1632,8 +1633,10 @@ private Flux<? extends PackageResource> getPackages(String applicationGuid) {
     }
 
     @Override
-    public List<UserRole> getUserRolesBySpaceAndUser(UUID spaceGuid, UUID userGuid) {
-        return fetchList(() -> getRoles(spaceGuid, userGuid), ImmutableRawUserRole::of);
+    public Set<UserRole> getUserRolesBySpaceAndUser(UUID spaceGuid, UUID userGuid) {
+        Supplier<Set<UserRole>> setFactory = () -> EnumSet.noneOf(UserRole.class);
+        return fetchFlux(() -> getRoles(spaceGuid, userGuid), ImmutableRawUserRole::of).collect(Collectors.toCollection(setFactory))
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:1460]: Cool

File: src/main/java/com/sap/cloudfoundry/client/facade/CloudControllerClient.java [hunk 2]
```diff
@@ -645,18 +529,6 @@ void createApplication(String applicationName, Staging staging, Integer disk, In
      */
     List<CloudStack> getStacks();
 
-    /**
-     * Login using the credentials already set for the client.
-     *
-     * @return authentication token
-     */
-    OAuth2AccessTokenWithAdditionalInfo login();
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/CloudControllerClient.java:653]: What is used in DataTerminationService and OrphanedDataCleaner when there is no login

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/ApplicationLogEntity.java [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 import java.util.Map;
 
+//TODO this won't be needed if we use the ReactorLogCacheClient
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/ApplicationLogEntity.java:9]: Complete the TODO :D

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java [hunk 4]
```diff
@@ -46,6 +52,28 @@ public LogCacheClient createLogCacheClient(URL controllerUrl, OAuthClient oAuthC
...
 
+    public CloudSpaceClient createSpaceClient(URL controllerUrl, OAuthClient oAuthClient, Map<String, String> requestTags) {
+        var spacesV3 = createV3SpacesClient(controllerUrl, oAuthClient, requestTags);
+        var orgsV3 = createV3OrgsClient(controllerUrl, oAuthClient, requestTags);
+        return new CloudSpaceClient(spacesV3, orgsV3);
+    }
+
+    private SpacesV3 createV3SpacesClient(URL controllerUrl, OAuthClient oAuthClient, Map<String, String> requestTags) {
+        //either use this hardcoded value or make a call to / (root) and get .links.cloud_controller_v3.href
+        // the root endpoint does not need authentication
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java:63]: Get the link from the "/" endpoint

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java [hunk 5]
```diff
@@ -46,6 +52,28 @@ public LogCacheClient createLogCacheClient(URL controllerUrl, OAuthClient oAuthC
...
+        String v3Api = controllerUrl.toString() + "/v3";
+        return new ReactorSpacesV3(getOrCreateConnectionContext(controllerUrl.getHost()), Mono.just(v3Api),
+                                   oAuthClient.getTokenProvider(), requestTags);
+    }
+
+    private OrganizationsV3 createV3OrgsClient(URL controllerUrl, OAuthClient oAuthClient, Map<String, String> requestTags) {
+        //either use this hardcoded value or make a call to / (root) and get .links.cloud_controller_v3.href
+        // the root endpoint does not need authentication
+        String v3Api = controllerUrl.toString() + "/v3";
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java:72]: Same.
What was the exact reason for using a custom client?

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 6]
```diff
@@ -8,27 +8,35 @@
 import java.net.http.HttpResponse;
 import java.net.http.HttpResponse.BodyHandlers;
 import java.nio.charset.StandardCharsets;
-import java.time.Duration;
-import java.time.LocalDateTime;
-import java.time.ZoneOffset;
+import java.time.*;
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:11]: No!

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 7]
```diff
@@ -8,27 +8,35 @@
...
 import com.sap.cloudfoundry.client.facade.Messages;
+import com.sap.cloudfoundry.client.facade.domain.ApplicationLog;
+import com.sap.cloudfoundry.client.facade.domain.ImmutableApplicationLog;
 import com.sap.cloudfoundry.client.facade.oauth2.OAuthClient;
 
 import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
 import org.apache.commons.io.IOUtils;
+import org.cloudfoundry.logcache.v1.*;
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:30]: No

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 8]
```diff
@@ -8,27 +8,35 @@
...
+import org.cloudfoundry.reactor.logcache.v1.ReactorLogCacheClient;
 import org.springframework.http.HttpHeaders;
 import org.springframework.http.HttpStatus;
 import org.springframework.web.util.UriComponentsBuilder;
-import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.util.retry.Retry;
+import reactor.util.retry.RetryBackoffSpec;
 
+//TODO run stress tests and observe direct memory with ReactorLogCacheClient
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:39]: Do not forget to run the tests and delete the comment

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 9]
```diff
@@ -37,13 +45,71 @@ public class LogCacheClient {
     private final OAuthClient oAuthClient;
     private final Map<String, String> requestTags;
 
+//    private final org.cloudfoundry.logcache.v1.LogCacheClient delegate;
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:48]: delete dis

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 10]
```diff
@@ -37,13 +45,71 @@ public class LogCacheClient {
...
+    private Throwable throwOriginalError(RetryBackoffSpec retrySpec, Retry.RetrySignal signal) {
+        return signal.failure();
+    }
+
+    public List<ApplicationLog> getRecentLogs0(UUID appGuid, LocalDateTime offset) {
+        org.cloudfoundry.logcache.v1.LogCacheClient client = ReactorLogCacheClient.builder()
+                .requestTags(requestTags)
+                .tokenProvider(oAuthClient.getTokenProvider())
+                .root(Mono.just(logCacheApi))
+//                .connectionContext(<>)
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:87]: Remove this

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 11]
```diff
@@ -37,13 +45,71 @@ public class LogCacheClient {
...
+    private static ApplicationLog.MessageType fromLogMessageType0(LogType logType) {
+        return logType == LogType.OUT ? ApplicationLog.MessageType.STDOUT : ApplicationLog.MessageType.STDERR;
+    }
+
+    private Throwable throwOriginalError(RetryBackoffSpec retrySpec, Retry.RetrySignal signal) {
+        return signal.failure();
+    }
+
+    public List<ApplicationLog> getRecentLogs0(UUID appGuid, LocalDateTime offset) {
+        org.cloudfoundry.logcache.v1.LogCacheClient client = ReactorLogCacheClient.builder()
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:83]: Use var here

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 12]
```diff
@@ -37,13 +45,71 @@ public class LogCacheClient {
...
+//    private final org.cloudfoundry.logcache.v1.LogCacheClient delegate;
+
     public LogCacheClient(String logCacheApi, OAuthClient oAuthClient, Map<String, String> requestTags) {
         this.logCacheApi = logCacheApi;
         this.oAuthClient = oAuthClient;
         this.requestTags = requestTags;
     }
 
-    public Flux<ApplicationLogEntity> getRecentLogs(UUID applicationGuid, LocalDateTime offset) {
+    private ApplicationLog mapToAppLog(Envelope envelope) {
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:56]: Sort the methods so they are chained and maybe rename to **mapEnvelopeTpAppLog**

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 13]
```diff
@@ -37,13 +45,71 @@ public class LogCacheClient {
...
+        var tags = envelope.getTags();
+        var log = envelope.getLog();
+        return ImmutableApplicationLog.builder()
+                                      .applicationGuid(envelope.getSourceId())
+                                      .message(log.getPayloadAsText())
+                                      .timestamp(fromLogTimestamp(envelope.getTimestamp()))
+                                      .messageType(fromLogMessageType0(log.getType()))
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:63]: Do not forget to rename if it works as expected

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    public CloudSpace getSpace(UUID spaceGuid) {
+        GetSpaceRequest spaceRequest = GetSpaceRequest.builder()
+                                                      .spaceId(spaceGuid.toString())
+                                                      .build();
+        Space space = spacesClient.get(spaceRequest)
+                                  .block();
+        if (space == null) {
+            throw new CloudOperationException(HttpStatus.NOT_FOUND, "Not Found");
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java:39]: Add logging what is not found.

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        GetOrganizationRequest orgRequest = GetOrganizationRequest.builder()
+                                                                  .organizationId(space.getRelationships()
+                                                                                       .getOrganization()
+                                                                                       .getData()
+                                                                                       .getId())
+                                                                  .build();
+        Organization org = orgsClient.get(orgRequest)
+                                     .block();
+        if (org == null) {
+            throw new CloudOperationException(HttpStatus.NOT_FOUND, "Not Found");
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java:51]: Add logging what is not found.

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        this.spacesClient = spacesClient;
+        this.orgsClient = orgsClient;
+    }
+
+    public CloudSpace getSpace(UUID spaceGuid) {
+        GetSpaceRequest spaceRequest = GetSpaceRequest.builder()
+                                                      .spaceId(spaceGuid.toString())
+                                                      .build();
+        Space space = spacesClient.get(spaceRequest)
+                                  .block();
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java:37]: Inline the space request and orgRequest

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    public CloudSpace getSpace(String organizationName, String spaceName, boolean required) {
+        var orgsResponse = orgsClient.list(ListOrganizationsRequest.builder()
+                                                                   .name(encodeAsQueryParam(organizationName))
+                                                                   .build())
+                                     .block();
+        List<? extends Organization> orgs = orgsResponse.getResources();
+        if (orgs.isEmpty()) {
+            if (required) {
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java:68]: inline this org.isEmpty && required

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+        Organization org = orgs.get(0);
+
+        var spacesResponse = spacesClient.list(ListSpacesRequest.builder()
+                                                                .organizationId(org.getId())
+                                                                .name(encodeAsQueryParam(spaceName))
+                                                                .build())
+                                         .block();
+        List<? extends Space> spaces = spacesResponse.getResources();
+        if (spaces.isEmpty()) {
+            if (required) {
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java:83]: same as above

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+                throw new CloudOperationException(HttpStatus.NOT_FOUND, "Not Found", "Space " + spaceName + "not found in organization " + organizationName);
+            }
+            return null;
+        }
+
+        Space space = spaces.get(0);
+        return mapToAdapter(space, org);
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java:90]: rename to mapToCloudSpace

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        return mapToAdapter(space, org);
+    }
+
+    public CloudSpace getSpace(String organizationName, String spaceName) {
+        return getSpace(organizationName, spaceName, true);
+    }
+
+    public CloudSpace getSpace(String organizationName, String spaceName, boolean required) {
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudSpaceClient.java:61]: Let's remove this flag, only makes the code harder to read and more error prone

File: src/main/java/com/sap/cloudfoundry/client/facade/CloudControllerClient.java (hunk 2 above)
Comment (radito3) [reply src/main/java/com/sap/cloudfoundry/client/facade/CloudControllerClient.java:653]: In `DataTerminationService` we use a custom controller client, which creates its `WebClient` with an initialized `OAuthClient` from the credentials passed to it.
In `OrphanedDataCleaner` we do a similar thing, where we create a `CloudSpaceClient` with a new, initialized `OAuthClient`

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java [hunk 21]
```diff
@@ -46,6 +55,40 @@ public LogCacheClient createLogCacheClient(URL controllerUrl, OAuthClient oAuthC
         return new LogCacheClient(logCacheApi, oAuthClient, requestTags);
     }
 
+    @SuppressWarnings("unchecked")
+    public CloudSpaceClient createSpaceClient(URL controllerUrl, OAuthClient oAuthClient, Map<String, String> requestTags) {
+        String v3Api;
+        var httpClient = java.net.http.HttpClient.newHttpClient();
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java:61]: Why not imported?

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java [hunk 22]
```diff
@@ -46,6 +55,40 @@ public LogCacheClient createLogCacheClient(URL controllerUrl, OAuthClient oAuthC
...
+                                                      .GET()
+                                                      .uri(controllerUrl.toURI())
+                                                      .build(), HttpResponse.BodyHandlers.ofString());
+            var map = JsonUtil.convertJsonToMap(response.body());
+            var links = (Map<String, Object>) map.get("links");
+            var ccv3 = (Map<String, Object>) links.get("cloud_controller_v3");
+            v3Api = (String) ccv3.get("href");
+        } catch (Exception ignored) {
+            v3Api = controllerUrl.toString() + "/v3";
+        }
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java:73]: Log the exception

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java [hunk 23]
```diff
@@ -46,6 +55,40 @@ public LogCacheClient createLogCacheClient(URL controllerUrl, OAuthClient oAuthC
...
     }
 
+    @SuppressWarnings("unchecked")
+    public CloudSpaceClient createSpaceClient(URL controllerUrl, OAuthClient oAuthClient, Map<String, String> requestTags) {
+        String v3Api;
+        var httpClient = java.net.http.HttpClient.newHttpClient();
+        try {
+            var response = httpClient.send(HttpRequest.newBuilder()
+                                                      .GET()
+                                                      .uri(controllerUrl.toURI())
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java:65]: Is there a retry logic for this client?

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java [hunk 24]
```diff
@@ -111,4 +186,27 @@ private ApplicationLogsResponse parseBody(InputStream is) {
...
+                                                                         .getMessageType()))
+                                      .sourceName(log.getTags()
+                                                     .get("source_type"))
+                                      .build();
+    }
+
+    private static String decodeLogPayload(String base64Encoded) {
+        var result = Base64.getDecoder()
+                .decode(base64Encoded.getBytes(StandardCharsets.UTF_8));
+        return new String(result, StandardCharsets.UTF_8);
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/LogCacheClient.java:206]: Format

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java [hunk 25]
```diff
@@ -1,25 +1,47 @@
...
-import reactor.netty.http.client.HttpClient;
+import reactor.core.publisher.Mono;
 
 @Value.Immutable
 public abstract class CloudFoundryClientFactory {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(CloudFoundryClientFactory.class);
+
+    static final HttpClient HTTP_CLIENT = HttpClient.newBuilder()
+                                                    .executor(Executors.newSingleThreadExecutor())
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java:40]: This does mean that every request/response will be handled by one thread right?

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java [hunk 26]
```diff
@@ -41,11 +63,64 @@ public CloudFoundryClient createClient(URL controllerUrl, OAuthClient oAuthClien
...
+        HttpResponse<String> response;
+        try {
+            var request = HttpRequest.newBuilder()
+                                     .GET()
+                                     .uri(controllerUrl.toURI())
+                                     .timeout(Duration.ofMinutes(5))
+                                     .build();
+            response = HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofString());
+        } catch (Exception e) {
+            LOGGER.warn(MessageFormat.format("Calling {0} failed with: {1}", controllerUrl.toString(), e.getMessage()), e);
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java:90]: Extract in Messages class

File: src/main/java/com/sap/cloudfoundry/client/facade/util/CloudUtil.java [hunk 27]
```diff
@@ -70,4 +79,25 @@ public static <T> T parse(Class<T> clazz, Object object) {
         return defaultValue;
     }
 
+    public static <T> T executeWithRetry(Supplier<T> operation) {
+        for (int i = 1; i < RETRIES; i++) {
+            try {
+                return operation.get();
+            } catch (Exception e) {
+                LOGGER.warn(MessageFormat.format("Retrying operation that failed with message: {0}", e.getMessage()), e);
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/CloudUtil.java:87]: Extract in messages

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java (hunk 25 above)
Comment (radito3) [reply src/main/java/com/sap/cloudfoundry/client/facade/adapters/CloudFoundryClientFactory.java:40]: Yes, it should be.

**Review by IvanBorislavovDimitrov — APPROVED** [review]

**Pull request merged**